### DISCUSSION TITLE
Add GUI button to trigger an action in a generic inventory

### DISCUSSION THREAD

**Discussion by Noitcereon** [FabricMC/fabric-api#2457] [Mod Dev Support] [answered]
Hello. I am trying to make a button that can trigger my sort function, but I can't seem to figure out how to do it.

I've succeded in adding a button, but I don't know how to hook it up with the correct data to trigger the sort.

### The button in-game:
![image](https://user-images.githubusercontent.com/40148361/183624367-fd26a30f-6e7c-47e0-8163-3ed548ea2e86.png)

### The code that added the button:
![image](https://user-images.githubusercontent.com/40148361/183624457-f816c26c-e00a-42b8-a6d3-47b3a12f3f2a.png)

### The code I need to trigger:
![image](https://user-images.githubusercontent.com/40148361/183624664-b84bb5cc-a9d0-4b9d-9b31-9e98bf19e10d.png)

So my question is then: 

**How do I add a GUI button to a chest/generic container that can trigger a function requiring the inventory of the container?**

---------------
### My Progress (See copiable code at bottom for solution)

I got it to work in the logical client, but it does not sync with the logical server:

![image](https://user-images.githubusercontent.com/40148361/183656481-70ff2276-28ae-44c3-a178-66c200c84b5a.png)
---------------
After trying to follow warjort's answer (through fabric wiki about networking and trial and error) I got it to work with the following:
This solution, however, does not work with a dedicated server.

![image](https://user-images.githubusercontent.com/40148361/184612962-d31ba7c6-0c79-451c-8114-a4622dfd48f9.png)
--------------- 

After moving the PlayNetworking.registerGlobalReceiver(...) to onInitialize in ModInitializer, moving ScreenEvents.register to Client initializer + some refactoring:

```
package com.noitcereon.simplesorting;

// imports removed for brevity

public class SimpleSortingMod implements ModInitializer {
	// This logger is used to write text to the console and the log file.
	// It is considered best practice to use your mod id as the logger's name.
	// That way, it's clear which mod wrote info, warnings, and errors.
	public static final Logger LOGGER = LoggerFactory.getLogger(ModInfo.MOD_ID);
	public static final Identifier INVENTORY_SORT_REQUEST_ID = new Identifier(ModInfo.MOD_ID, "inventory-sort-request");

	@Override
	public void onInitialize() {
		// This code runs as soon as Minecraft is in a mod-load-ready state.
		// However, some things (like resources) may still be uninitialized.
		// Proceed with mild caution.
		LOGGER.info("Initialising {}", ModInfo.MOD_ID);

		ServerPlayNetworking.registerGlobalReceiver(INVENTORY_SORT_REQUEST_ID, (server, player, handler, buf, responseSender) -> {
			server.execute(() -> {
				LOGGER.debug("Server received package: " + INVENTORY_SORT_REQUEST_ID.getNamespace() + " " + INVENTORY_SORT_REQUEST_ID.getPath());

				sortCurrentlyOpenInventory(player);
			});
		});
	}

	private void sortCurrentlyOpenInventory(ServerPlayerEntity player) {
		ScreenHandler screenHandler = player.currentScreenHandler;
            if (screenHandler instanceof GenericContainerScreenHandler containerScreenHandler) {
                if (!screenHandler.canUse(player)){
                    LOGGER.info("Failed to sort, because player cannot use the container anymore.");
                    return;
                }
    
                Inventory containerInventory = containerScreenHandler.getInventory();
                InventorySorter.sortInventory(containerInventory);
                containerInventory.markDirty();
            }
        else{
            LOGGER.warn("player.currentScreenHandler returned {}. Simply Sorting uses GenericContainerScreenHandler to function.", screenHandler.getClass());
        }
	}
}
```
AND
```
public class SimpleSortingModClient implements ClientModInitializer {

    @Override
    public void onInitializeClient() {
        ScreenEvents.AFTER_INIT.register((client, screen, scaledWidth, scaledHeight) -> {
            if (screen instanceof GenericContainerScreen) {
                Screens.getButtons(screen).add(
                        new ButtonWidget((scaledWidth / 100) * 70, (scaledHeight / 100) * 5, 40, 16, Text.literal("Sort"), (btn) -> {
                            SimpleSortingMod.LOGGER.debug("Sort Button pressed");
                            PacketByteBuf packet = PacketByteBufs.empty();
                            ClientPlayNetworking.send(SimpleSortingMod.INVENTORY_SORT_REQUEST_ID, packet); // this triggers the global receiver registered with the same identifier.
                        })
                );
            }
        });
    }
}

**Comment by warjort** [comment, marked as answer]
You don't show all your code, but I think you are mixing together code from the client/server.
e.g. that syncState() is used by the server to send slot updates from the server to the client

What you are doing is not a usecase I'm familiar with, but...

I would guess what you really need to do is have your button send a network request to the server
then use player.currentScreenHandler to get the ScreenHandler and do all the work there.

**Reply by Noitcereon** [reply]
Thanks for the answer, I'll take a look at it when I have time. Something else popped up, which I have to prioritize over this mod.

**Reply by Noitcereon** [reply]
I tried to do what you said and it seems to work:

![image](https://user-images.githubusercontent.com/40148361/184612962-d31ba7c6-0c79-451c-8114-a4622dfd48f9.png)

**Reply by warjort** [reply]
You don't want to do it like this.

e.g. what happens if the client and server are on different machines? 
There will no be receiver on the remote server to handle your network request.
You are only registering the receiver when the user clicks the button, which never happens on a dedicated server.
You want to register it in your mod initialisation.

Your reference to "client" in the handler should be "server". 
You want your code to run on the server thread (not the client thread).

**Reply by Noitcereon** [reply]
Alright, I presume this is what you mean then:

```
package com.noitcereon.simplesorting;

// imports removed for brevity

public class SimpleSortingMod implements ModInitializer {
	// This logger is used to write text to the console and the log file.
	// It is considered best practice to use your mod id as the logger's name.
	// That way, it's clear which mod wrote info, warnings, and errors.
	public static final Logger LOGGER = LoggerFactory.getLogger(ModInfo.MOD_ID);
	public static final Identifier INVENTORY_SORT_REQUEST_ID = new Identifier(ModInfo.MOD_ID, "inventory-sort-request");

	@Override
	public void onInitialize() {
		// This code runs as soon as Minecraft is in a mod-load-ready state.
		// However, some things (like resources) may still be uninitialized.
		// Proceed with mild caution.
		LOGGER.info("Initialising {}", ModInfo.MOD_ID);

		ServerPlayNetworking.registerGlobalReceiver(INVENTORY_SORT_REQUEST_ID, (server, player, handler, buf, responseSender) -> {
			server.execute(() -> {
				LOGGER.debug("Server received package: " + INVENTORY_SORT_REQUEST_ID.getNamespace() + " " + INVENTORY_SORT_REQUEST_ID.getPath());

				sortCurrentlyOpenInventory(player);
			});
		});

		ScreenEvents.AFTER_INIT.register((client, screen, scaledWidth, scaledHeight) -> {
			if (screen instanceof GenericContainerScreen) {
				Screens.getButtons(screen).add(
						new ButtonWidget((scaledWidth / 100) * 70, (scaledHeight / 100) * 5, 40, 16, Text.literal("Sort"), (btn) -> {
							LOGGER.debug("Sort Button pressed");
							PacketByteBuf packet = PacketByteBufs.empty();
							ClientPlayNetworking.send(INVENTORY_SORT_REQUEST_ID, packet); // this triggers the global receiver registered with the same identifier.
						})
				);
			}
		});
	}
	private void sortCurrentlyOpenInventory(ServerPlayerEntity player) {
		GenericContainerScreenHandler screenHandler = (GenericContainerScreenHandler) player.currentScreenHandler;
		Inventory containerInventory = screenHandler.getInventory();
		InventorySorter.sortInventory(containerInventory);
		containerInventory.markDirty();
	}
}

```

As image with color coding: 
![image](https://user-images.githubusercontent.com/40148361/184643278-5482c410-779e-4915-a070-f528cc0f7f91.png)

**Reply by warjort** [reply]
Yes. 
Although like I said above, I don't know your usecase so I can't comment on the actual implementation of the handler.

If it were me I would:
* Actually check the handler is the correct type before casting and log a warning instead of crashing
* call screenHandler.canUse(player) to check the block still exists, e.g. it hasn't been destroyed by an explosion during your processing :-)

Also your ScreenEvents register() call needs to be in a ClientModInitializer otherwise this will also crash on a dedicated server.

**Reply by Noitcereon** [reply]
My use case is to sort the inventory of a container, so it's easier to find the items you're looking for. At a basic level, just to make items of the same type to stick together. The Minecraft GenericScreenHandler makes this work for chests and barrels at least. It is not meant to be a comprehensive sorting mod.

I've added some safety checks to the sortCurrentlyOpenInventory:

![image](https://user-images.githubusercontent.com/40148361/184660231-62716c14-9d50-44b5-b8f9-37453939c2cd.png)

And moved the ScreenEvents register() call to SimpleSortingModClient which implements the ClientModInitalizer.

![image](https://user-images.githubusercontent.com/40148361/184660458-71ceca89-617b-4452-b4f6-12d745fe39eb.png)

Thanks for your help, it has been tremendously helpful. Especially since this is my first mod :)

My tiny project is here: https://github.com/Noitcereon/simple-sorting if you're interested in browsing the code.

**Reply by warjort** [reply]
You should do the cast like this:
````
if (screenHandler instanceof GenericContainerScreenHandler containerScreenHandler) {
    // Your code here
} else {
   LOGGER.warn("blah blah");
}
`````
I only suggested doing this in case some other mod changes player.currentScreenHandler in unexpected/weird ways.
It would mean your sort stops working if this happens, but at least the server doesn't crash. :-)

**Reply by Noitcereon** [reply]
Done:
![image](https://user-images.githubusercontent.com/40148361/184667500-9373bb18-b003-4762-997f-7361f76e03e1.png)

**Reply by warjort** [reply]
Ok, but that screenHandler might be null if we are handling weird behaviour, 
you shouldn't call getClass() on it without checking.

e.g. replace screenHandler.getClass() with
````
screenHandler == null ? "Null" : screenHandler.getClass().getName()
````
